lib/TopStoriesFeed.jsm

```diff
@@ -30,17 +30,29 @@ const SPOC_IMPRESSION_TRACKING_PREF = "feeds.section.topstories.spoc.impressions
 const REC_IMPRESSION_TRACKING_PREF = "feeds.section.topstories.rec.impressions";
 const OPTIONS_PREF = "feeds.section.topstories.options";
 const MAX_LIFETIME_CAP = 500; // Guard against misconfiguration on the server
+const DISCOVERY_STREAM_PREF = "discoverystream.config";
 
 this.TopStoriesFeed = class TopStoriesFeed {
-  constructor() {
-    this.spocCampaignMap = new Map();
+  constructor(ds) {
+    this.discoveryStreamEnabled = ds && ds.value && JSON.parse(ds.value).enabled;
+    if (!this.discoveryStreamEnabled) {
+      this.initializeProperties();
+    }
+  }
+
+  initializeProperties() {
     this.contentUpdateQueue = [];
+    this.spocCampaignMap = new Map();
     this.cache = new PersistentCache(SECTION_ID, true);
     this._prefs = new Prefs();
   }
 
   async onInit() {
     SectionsManager.enableSection(SECTION_ID);
+    if (this.discoveryStreamEnabled) {
+      return;
```

Comment:
We `Services.obs.addObserver()` further down which we skip. So we should check enabled before we try to remove in uninit:
```
```console.log: "action type" "UNINIT"
JavaScript error: resource://activity-stream/lib/TopStoriesFeed.jsm, line 114: NS_ERROR_ILLEGAL_VALUE: Component returned failure code: 0x80070057 (NS_ERROR_ILLEGAL_VALUE) [nsIObserverService.removeObserver]```

Reply:
Checking discovery stream enabled before invoking remove observer is not full proof. It leaves out edge case where user manually disable and enable discovery stream experience back.

I have updated to invoke removeObserver inside try catch so that it continue uninit after catching error. Not sure if there is a way to check if observer exists before invoking remove, that should definitely be a better handling.